### PR TITLE
Inji-400: Cache api calls and return initial values if network not available

### PR DESCRIPTION
**Pull request by tilak-puli**

### CODE REVIEW

File: shared/InitialConfig.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Initial configs used by app if network is not available when app is opened for first time. once network is available response is cached and used from then
+// Note: Need to keep this config in sync with actual mimoto response.
```
Comment (swatigoel) [inline shared/InitialConfig.ts:2]: add configuration file name here
In mosip-config, refer inji-default.properties

File: shared/InitialConfig.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    audience: 'ida-binding',
+    allowedInternalAuthType: 'otp,bio-Finger,bio-Iris,bio-Face',
+    vcDownloadMaxRetry: '10',
+    minStorageRequiredForAuditEntry: '2',
+    minStorageRequired: '2',
+    vcDownloadPoolInterval: '6000',
+    issuer: 'residentapp',
+    allowedAuthType: 'demo,otp,bio-Finger,bio-Iris,bio-Face',
+    allowedEkycAuthType: 'demo,otp,bio-Finger,bio-Iris,bio-Face',
+    warningDomainName: 'https://api.qa-inji.mosip.net',
```
Comment (swatigoel) [inline shared/InitialConfig.ts:16]: Remove warningDomainName and faceSdkModelUrl from caching as it depends on api response.
Let app handles these errors if these values are missing

File: shared/api.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+};
+
+interface cacheAPIMethodProps {
+  isCachePreferred?: boolean;
+  cacheKey: string;
+  fetchCall: (...props: any) => Promise<any>;
+  onErrorHardCodedValue?: any;
+}
+
+async function catchAPIMethod({
```
Comment (swatigoel) [inline shared/api.ts:82]: Do we need to use keyword `method` in function name?
and it's catchAPI or cacheAPI?

File: shared/api.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      const response = await fetchCall();
+
+      Storage.setItem(cacheKey, JSON.stringify(response)).then(() =>
+        console.log('Cached response for ' + cacheKey),
+      );
+
+      return response;
+    }
+  } catch (error) {
+    console.warn('Failed to load due to network issue');
```
Comment (swatigoel) [inline shared/api.ts:123]: Add function name in warning message which will help to understand this error has come for caching api

File: shared/api.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  } else {
+    return await cacheAPIMethodWithAPIPreference(
+      cacheKey,
+      fetchCall,
+      onErrorHardCodedValue,
+    );
+  }
+}
+
+async function cacheAPIMethodWithCachePreference(
```
Comment (swatigoel) [inline shared/api.ts:103]: Same here do we need to use keyword method in function name?

File: shared/api.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    if (onErrorHardCodedValue != undefined) {
+      return onErrorHardCodedValue;
+    } else {
+      throw error;
+    }
+  }
+}
+
+async function cacheAPIMethodWithAPIPreference(
```
Comment (swatigoel) [inline shared/api.ts:134]: Same here do we need to use keyword method in function name?

File: shared/api.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  try {
+    const response = (await Storage.getItem(cacheKey)) as string;
+
+    if (response) {
+      return JSON.parse(response);
+    } else {
+      const response = await fetchCall();
+
+      Storage.setItem(cacheKey, JSON.stringify(response)).then(() =>
+        console.log('Cached response for ' + cacheKey),
```
Comment (swatigoel) [inline shared/api.ts:117]: This should be logged in debugging mode only

File: shared/api.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+async function cacheAPIMethodWithAPIPreference(
+  cacheKey: string,
+  fetchCall: (...props: any[]) => any,
+  onErrorHardCodedValue?: any,
+) {
+  try {
+    const response = await fetchCall();
+    Storage.setItem(cacheKey, JSON.stringify(response)).then(() =>
+      console.log('Cached response for ' + cacheKey),
```
Comment (swatigoel) [inline shared/api.ts:142]: This should be logged in debugging mode only

**Review by swatigoel — CHANGES_REQUESTED** [review]

File: shared/api.ts (hunk 7 above)
Comment (tw-tilak-puli) [reply shared/api.ts:117]: All logs are only visible in debug mode

File: shared/InitialConfig.ts (hunk 1 above)
Comment (tilak-puli) [reply shared/InitialConfig.ts:2]: sure. adding

File: shared/InitialConfig.ts (hunk 2 above)
Comment (tilak-puli) [reply shared/InitialConfig.ts:17]: removed

File: shared/api.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  } else {
+    return await generateCacheAPIFunctionWithAPIPreference(
+      cacheKey,
+      fetchCall,
+      onErrorHardCodedValue,
+    );
+  }
+}
+
+async function generateCacheAPIFunctionWithCachePreference(
```
Comment (swatigoel) [inline shared/api.ts:103]: This is the actual method which use cached response or fetch data from api.
We can name it as cacheAPIWithCachePreference

**Review by swatigoel — CHANGES_REQUESTED** [review]

File: shared/InitialConfig.ts (hunk 2 above)
Comment (tilak-puli) [reply shared/InitialConfig.ts]: removed now. after removing also app works fine

File: shared/InitialConfig.ts (hunk 1 above)
Comment (tilak-puli) [reply shared/InitialConfig.ts:2]: added near all properties key

**Review by swatigoel — APPROVED** [review]

**Pull request merged**